src/main/scala/is/hail/io/bgen/BgenRecord.scala

```diff
@@ -0,0 +1,176 @@
+              new DosageGenotype(gt, px)
+            } else {
+              setWarning(dosageGreaterThanTolerance)
+              noCall
+            }
+          } else {
+            if (dsum == 0)
+              setWarning(dosageNoCall)
+            else
+              setWarning(dosageLessThanTolerance)
+
+            noCall
+          }
+        }
+      }
+    }
+  }
+}
+
+class BgenRecordV12(compressed: Boolean, nSamples: Int, tolerance: Double) extends BgenRecord {
+  var expectedDataSize: Int = _
+  var expectedNumAlleles: Int = _
+
+  def setExpectedDataSize(size: Int) {
+    this.expectedDataSize = size
+  }
+
+  def setExpectedNumAlleles(n: Int) {
+    this.expectedNumAlleles = n
+  }
+
+  override def getValue: Iterable[Genotype] = {
+    require(input != null, "called getValue before serialized value was set")
+
+    val bytes = if (compressed) decompress(input, expectedDataSize) else input
+    val nRow = (bytes(0) & 0xff) | ((bytes(1) & 0xff) << 8) | ((bytes(2) & 0xff) << 16) | ((bytes(3) & 0xff) << 24)
+    assert(nRow == nSamples, "row nSamples is not equal to header nSamples")
+
+    val nAlleles = (bytes(4) & 0xff) | ((bytes(5) & 0xff) << 8)
+    assert(nAlleles == expectedNumAlleles, s"Value for `nAlleles' in genotype probability data storage is not equal to value in variant identifying data. Expected $expectedNumAlleles but found $nAlleles.")
+
+    val minPloidy = bytes(6) & 0xff
+    val maxPloidy = bytes(7) & 0xff
+
+    if (minPloidy != 2 || maxPloidy != 2)
+      fatal(s"Hail only supports diploid genotypes. Found min ploidy equals `$minPloidy' and max ploidy equals `$maxPloidy'.")
+
+    val samplePloidy = Array.ofDim[Int](nSamples)
+    var i = 0
+    while (i < nSamples) {
+      val b = bytes(i + 8)
+      val isMissing = ((b & 0xffL) >>> 7) == 1
+      val ploidy = if (isMissing) -1 else b & 0x3f
+      assert(ploidy >= -1 && ploidy <= 63, s"Ploidy value must be in range [0, 63]. Found $ploidy.")
+      samplePloidy(i) = ploidy
+      i += 1
+    }
+    assert(i == nSamples, s"Number of ploidy values `$i' does not equal the number of samples `$nSamples'.")
+
+    val phase = bytes(nSamples + 8) & 0xff
+    assert(phase == 0 || phase == 1, s"Value for phase must be 0 or 1. Found $phase.")
+    val isPhased = phase == 1
+
+    if (isPhased)
+      fatal("Hail does not support phased genotypes.")
+
+    val nBitsPerProb = bytes(nSamples + 9) & 0xff
+    assert(nBitsPerProb >= 1 && nBitsPerProb <= 32, s"Value for nBits must be between 1 and 32 inclusive. Found $nBitsPerProb.")
+
+    val probabilityIterator = new BgenProbabilityIterator(bytes, nBitsPerProb, nSamples + 10)
+
+    val nGenotypes = triangle(nAlleles)
+    val sampleProbs = Array.ofDim[Long](nGenotypes)
+    val totalProbInt = ~0L >>> (64 - nBitsPerProb)
+    val noCall: Genotype = new DosageGenotype(-1, null)
+
+    val nExpectedBytesProbs = math.ceil((nSamples * (nGenotypes - 1) * nBitsPerProb).toDouble / 8).toInt
```

Comment:
You don't need Double math.  Write `nSamples *  (nGenotypes - 1) * nBitsPerProb + 7) / 8`.  `... + 7) / 8` has the effect of rounding up.